---
title: "Lab 5"
author: "Jess Roy"
date: "2/19/2020"
output: html_document
---


```{eval = FALSE}
 download.file(url="https://ndownloader.figshare.com/files/2292169",
              destfile = "data/portal_data_joined.csv"
              
View(surveys)
```

Data Carpentry:
```{r}
library (tidyverse)
surveys <- read_csv("portal_data_joined.csv")
```

```{r}
## inspect the data
str(surveys)
```


Selecting:
To select columns of a data frame, use select(). The first argument to this function is the data frame (surveys), and the subsequent arguments are the columns to keep
```{r}
select(surveys, plot_id, species_id, weight)
```

To select all columns except certain ones, put a “-” in front of the variable to exclude it.
```{r}
select(surveys, -record_id, -species_id)
```

To choose rows based on a specific criteria, use filter():
```{r}
filter(surveys, year == 1995)
```

Pipes:
What if you want to select and filter at the same time? There are three ways to do this: use intermediate steps, nested functions, or pipes.
With intermediate steps, you create a temporary data frame and use that as input to the next function, like this:
```{r}
surveys2 <- filter(surveys, weight < 5)
surveys_sml <- select(surveys2, species_id, sex, weight)
```

You can also nest functions (i.e. one function inside of another), like this:
```{r}
surveys_sml <- select(filter(surveys, weight < 5), species_id, sex, weight)
```

The last option, pipes, are a recent addition to R. Pipes let you take the output of one function and send it directly to the next, which is useful when you need to do many things to the same dataset. Pipes in R look like %>% and are made available via the magrittr package, installed automatically with dplyr. If you use RStudio, you can type the pipe with Ctrl + Shift + M if you have a PC or Cmd + Shift + M if you have a Mac.
```{r}
surveys %>%
  filter(weight < 5) %>%
  select(species_id, sex, weight)
```

If we want to create a new object with this smaller version of the data, we can assign it a new name:
```{r}
surveys_sml <- surveys %>%
  filter(weight < 5) %>%
  select(species_id, sex, weight)

surveys_sml
```

Challenge
Using pipes, subset the surveys data to include animals collected before 1995 and retain only the columns year, sex, and weight.
```{r}
surveys %>%
    filter(year < 1995) %>%
    select(year, sex, weight)
```

Mutate
Frequently you’ll want to create new columns based on the values in existing columns, for example to do unit conversions, or to find the ratio of values in two columns. For this we’ll use mutate().

To create a new column of weight in kg:
```{r}
surveys %>%
  mutate(weight_kg = weight / 1000)
```

You can also create a second new column based on the first new column within the same call of mutate():
```{r}
surveys %>%
  mutate(weight_kg = weight / 1000,
         weight_lb = weight_kg * 2.2)
```

If this runs off your screen and you just want to see the first few rows, you can use a pipe to view the head() of the data. (Pipes work with non-dplyr functions, too, as long as the dplyr or magrittr package is loaded).
```{r}
surveys %>%
  mutate(weight_kg = weight / 1000) %>%
  head()
```

The first few rows of the output are full of NAs, so if we wanted to remove those we could insert a filter() in the chain:
```{r}
surveys %>%
  filter(!is.na(weight)) %>%
  mutate(weight_kg = weight / 1000) %>%
  head()
```

Challenge
Create a new data frame from the surveys data that meets the following criteria: contains only the species_id column and a new column called hindfoot_half containing values that are half the hindfoot_length values. In this hindfoot_half column, there are no NAs and all values are less than 30.
```{r}
surveys_hindfoot_half <- surveys %>%
    filter(!is.na(hindfoot_length)) %>%
    mutate(hindfoot_half = hindfoot_length / 2) %>%
    filter(hindfoot_half < 30) %>%
    select(species_id, hindfoot_half)
```

Split-apply-combine data analysis and the summarize() function
Many data analysis tasks can be approached using the split-apply-combine paradigm: split the data into groups, apply some analysis to each group, and then combine the results. dplyr makes this very easy through the use of the group_by() function.

The summarize() function
group_by() is often used together with summarize(), which collapses each group into a single-row summary of that group. group_by() takes as arguments the column names that contain the categorical variables for which you want to calculate the summary statistics. So to compute the mean weight by sex:
```{r}
surveys %>%
  group_by(sex) %>%
  summarize(mean_weight = mean(weight, na.rm = TRUE))
```

You can also group by multiple columns:
```{r}
surveys %>%
  group_by(sex, species_id) %>%
  summarize(mean_weight = mean(weight, na.rm = TRUE))
```

When grouping both by sex and species_id, the last few rows are for animals that escaped before their sex and body weights could be determined. You may notice that the last column does not contain NA but NaN (which refers to “Not a Number”). To avoid this, we can remove the missing values for weight before we attempt to calculate the summary statistics on weight. Because the missing values are removed first, we can omit na.rm = TRUE when computing the mean:
```{r}
surveys %>%
  filter(!is.na(weight)) %>%
  group_by(sex, species_id) %>%
  summarize(mean_weight = mean(weight))
```

If you want to display more data, you can use the print() function at the end of your chain with the argument n specifying the number of rows to display:
```{r}
surveys %>%
  filter(!is.na(weight)) %>%
  group_by(sex, species_id) %>%
  summarize(mean_weight = mean(weight)) %>%
  print(n = 15)
```

Once the data are grouped, you can also summarize multiple variables at the same time (and not necessarily on the same variable). For instance, we could add a column indicating the minimum weight for each species for each sex:
```{r}
surveys %>%
  filter(!is.na(weight)) %>%
  group_by(sex, species_id) %>%
  summarize(mean_weight = mean(weight),
            min_weight = min(weight))
```

It is sometimes useful to rearrange the result of a query to inspect the values. For instance, we can sort on min_weight to put the lighter species first:
```{r}
surveys %>%
  filter(!is.na(weight)) %>%
  group_by(sex, species_id) %>%
  summarize(mean_weight = mean(weight),
            min_weight = min(weight)) %>%
  arrange(min_weight)
```

To sort in descending order, we need to add the desc() function. If we want to sort the results by decreasing order of mean weight:
```{r}
surveys %>%
  filter(!is.na(weight)) %>%
  group_by(sex, species_id) %>%
  summarize(mean_weight = mean(weight),
            min_weight = min(weight)) %>%
  arrange(desc(mean_weight))
```

Counting
When working with data, we often want to know the number of observations found for each factor or combination of factors. For this task, dplyr provides count(). For example, if we wanted to count the number of rows of data for each sex, we would do:
```{r}
surveys %>%
    count(sex) 
```

The count() function is shorthand for something we’ve already seen: grouping by a variable, and summarizing it by counting the number of observations in that group. In other words, surveys %>% count() is equivalent to
```{r}
surveys %>%
    group_by(sex) %>%
    summarise(count = n())
```

For convenience, count() provides the sort argument:
```{r}
surveys %>%
    count(sex, sort = TRUE) 
```

If we wanted to count combination of factors, such as sex and species, we would specify the first and the second factor as the arguments of count():
```{r}
surveys %>%
  count(sex, species) 
```

With the above code, we can proceed with arrange() to sort the table according to a number of criteria so that we have a better comparison. For instance, we might want to arrange the table above in (i) an alphabetical order of the levels of the species and (ii) in descending order of the count:
```{r}
surveys %>%
  count(sex, species) %>%
  arrange(species, desc(n))
```

Challenge:
1.How many animals were caught in each plot_type surveyed?
```{r}
surveys %>%
    count(plot_type) 
```

2.Use group_by() and summarize() to find the mean, min, and max hindfoot length for each species (using species_id). Also add the number of observations (hint: see ?n).
```{r}
surveys %>%
    filter(!is.na(hindfoot_length)) %>%
    group_by(species_id) %>%
    summarize(
        mean_hindfoot_length = mean(hindfoot_length),
        min_hindfoot_length = min(hindfoot_length),
        max_hindfoot_length = max(hindfoot_length),
        n = n()
    )
```

3. What was the heaviest animal measured in each year? Return the columns year, genus, species_id, and weight.
```{r}
surveys %>%
    filter(!is.na(weight)) %>%
    group_by(year) %>%
    filter(weight == max(weight)) %>%
    select(year, genus, species, weight) %>%
    arrange(year)
```

Exporting data
Let’s start by removing observations of animals for which weight and hindfoot_length are missing, or the sex has not been determined:
```{r}
surveys_complete <- surveys %>%
  filter(!is.na(weight),           # remove missing weight
         !is.na(hindfoot_length),  # remove missing hindfoot_length
         !is.na(sex))                # remove missing sex
```

Because we are interested in plotting how species abundances have changed through time, we are also going to remove observations for rare species (i.e., that have been observed less than 50 times). We will do this in two steps: first we are going to create a data set that counts how often each species has been observed, and filter out the rare species; then, we will extract only the observations for these more common species:
```{r}
## Extract the most common species_id
species_counts <- surveys_complete %>%
    count(species_id) %>% 
    filter(n >= 50)

## Only keep the most common species
surveys_complete <- surveys_complete %>%
  filter(species_id %in% species_counts$species_id)
```

Now that our data set is ready, we can save it as a CSV file in our data folder.
```{r}
write_csv(surveys_complete, path = "surveys_complete.csv")
```


Exercises:

E1
```{r}
library (tidyverse)
SNPs <- read_tsv("23andMe_complete.txt", skip = 14, col_types = 
  cols(chromosome = col_factor()))
```

E2
```{r}
select(filter(SNPs, chromosome == "MT"), genotype)
```

E3
```{r}
SNPs %>%
  group_by(chromosome) %>%
  summarize(min_position = min(position),
            max_position = sum(position),
            total_position = sum(position))
```

E4
```{r}
SNPs2 <- SNPs %>%
  mutate(density = sum(position)/max(position)-min(position))
```

E5
```{r}
SNPs2 %>%
arrange(desc(density))
```

E6
```{r}
write_csv(SNPs2, path = "23andMe_SNPs2.csv")
```

